elfi/methods/diagnostics.py

```diff
@@ -0,0 +1,260 @@
+            E_ss = self._calc_entropy(thetas_ss, n_acc, k)
+            if E_ss < E_me:
+                E_me = E_ss
+                names_ss_me = names_ss
+                thetas_closest = thetas_ss[:n_closest]
+            logger.info('Combination %s shows the entropy of %f' % (names_ss, E_ss))
+        logger.info('\nThe minimum entropy of %f was found in %s.\n' % (E_me, names_ss_me))
+
+        # Find the summary-statistics combination with
+        # the minimum mean root sum of squared error (MRRSE).
+        MRSSE_min = np.inf
+        for set_ss in sets_candidate_ss:
+            names_ss = [ss.__name__ for ss in set_ss]
+            MRSSE_ss = self._calc_MRSSE(set_ss, thetas_closest, thetas[set_ss])
+            if MRSSE_ss < MRSSE_min:
+                MRSSE_min = MRSSE_ss
+                names_ss_MRSSE = names_ss
+                set_ss_2stage = set_ss
+            logger.info('Combination %s shows the MRSSE of %f' % (names_ss, MRSSE_ss))
+        logger.info('\nThe minimum MRSSE of %f was found in %s.' % (MRSSE_min, names_ss_MRSSE))
+        return set_ss_2stage
+
+    def _create_ss_combinations(self, max_cardinality):
+        """Create all combinations of the initialised summary statistics up till the maximum cardinality.
+
+        Parameters
+        ----------
+        max_cardinality : int
+            Cardinality limit of a summary-statistics combination.
+
+        Returns
+        -------
+        array_like
+            Combinations of summary statistics.
+
+        """
+        if max_cardinality > len(self.list_ss):
+            max_cardinality = len(self.list_ss)
+
+        combinations_ss = []
+        for i in range(max_cardinality):
+            for combination in combinations(self.list_ss, i + 1):
+                combinations_ss.append(combination)
+        return combinations_ss
+
+    def _obtain_accepted_thetas(self, set_ss, n_sim, n_acc):
+        """Obtain the parameters accepted by the ABC-rejection sampling.
+
+        The sampling is performed using the initialised simulator.
+
+        Parameters
+        ----------
+        set_ss : array_like
+            Combination of the summary statistics used in the rejection sampling.
+        n_sim : int
+            Number of the iterations of the rejection sampling.
+        n_acc : int
+            Number of the accepted parameters.
+
+        Returns
+        -------
+        array_like
+            Accepted parameters.
+
+        """
+        # Initialise the distance function.
+        list_ss = []
+        for ss in set_ss:
+            list_ss.append(elfi.Summary(ss, self.simulator))
+        if isinstance(self.fn_distance, str):
+            d = elfi.Distance(self.fn_distance, *list_ss)
+        else:
+            d = elfi.Discrepancy(self.fn_distance, *list_ss)
+
+        # Run the simulations.
+        sampler_rejection = elfi.Rejection(d, batch_size=1, seed=self.seed)
+        result = sampler_rejection.sample(n_acc, n_sim=n_sim)
```

Comment:
Hmm... I don't think it's necessary to rerun the simulator all the time. You could use a pool here to reuse precomputed results.

Reply:
The rationale of the selection is based on obtaining the simulations based on different summary-statistics combinations. The simulations corresponding to different combinations are also preseverd, as these are assessed in Stage 2: the parameters of the simulations of each summary-statistics combination are compared to the 'closest' parameters of the simulations of the minimum-entropy summary-statistics combination.

Reply:
I mean you could use a pool to cache the outputs from the simulator. ELFI would then here just recalculate what is needed, i.e. the summary statistics and distance.

Reply:
In general, the simulator may be computationally very heavy and you really want to avoid running it unless absolutely necessary.